commit 3bb748257405e94e13de76573a4e9da1cfd961d0
Author: Florian Weimer <fweimer@redhat.com>
Date:   Tue Jul 3 15:54:49 2018 +0200

    libio: Disable vtable validation in case of interposition [BZ #23313]
    
    (cherry picked from commit c402355dfa7807b8e0adb27c009135a7e2b9f1b0)

Backported from the upstream release/2.27/master branch.

diff --git a/libio/vtables.c b/libio/vtables.c
index e364ea03edbfa75b..d6478e4bab9050ce 100644
--- a/libio/vtables.c
+++ b/libio/vtables.c
@@ -68,3 +68,19 @@ _IO_vtable_check (void)
 
   __libc_fatal ("Fatal error: glibc detected an invalid stdio handle\n");
 }
+
+/* Some variants of libstdc++ interpose _IO_2_1_stdin_ etc. and
+   install their own vtables directly, without calling _IO_init or
+   other functions.  Detect this by looking at the vtables values
+   during startup, and disable vtable validation in this case.  */
+#ifdef SHARED
+__attribute__ ((constructor))
+static void
+check_stdfiles_vtables (void)
+{
+  if (_IO_2_1_stdin_.vtable != &_IO_file_jumps
+      || _IO_2_1_stdout_.vtable != &_IO_file_jumps
+      || _IO_2_1_stderr_.vtable != &_IO_file_jumps)
+    IO_set_accept_foreign_vtables (&_IO_vtable_check);
+}
+#endif
